File: Python/event_emitter_utils/event_emitter_utils_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
AllToolkit - Event Emitter Utilities Test Suite
================================================
Comprehensive test suite for the event emitter module.
"""

import sys
import os
import time
import threading
import asyncio

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mod import (
    EventEmitter, EventBus, Listener, EventRecord,
    create_emitter, create_channel,
    emits_event, on_event
)


class TestRunner:
    """Simple test runner with reporting."""
    
    def __init__(self):
        self.passed = 0
        self.failed = 0
        self.errors = []
    
    def test(self, name: str, condition: bool, error_msg: str = ""):
        """Run a single test."""
        if condition:
            self.passed += 1
            print(f"  ✓ {name}")
        else:
            self.failed += 1
            self.errors.append((name, error_msg))
            print(f"  ✗ {name}")
            if error_msg:
                print(f"    Error: {error_msg}")
    
    def report(self) -> bool:
        """Print test report and return success status."""
        total = self.passed + self.failed
        print(f"\n{'='*60}")
        print(f"Tests: {total} | Passed: {self.passed} | Failed: {self.failed}")
        
        if self.errors:
            print(f"\nFailed tests:")
            for name, msg in self.errors:
                print(f"  - {name}: {msg}")
        
        if self.failed == 0:
            print("✓ All tests passed!")
        else:
            print(f"✗ {self.failed} test(s) failed.")
        print('='*60)
        return self.failed == 0


def run_basic_tests(runner: TestRunner):
    """Test basic event emission and listening."""
    print("\nBasic Event Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    received_data = []
    
    def handler(data):
        received_data.append(data)
    
    # Test listener registration
    listener = emitter.on("test.event", handler)
    runner.test("on() returns Listener object", isinstance(listener, Listener))
    runner.test("Listener has correct callback", listener.callback == handler)
    
    # Test event emission
    emitter.emit("test.event", {"key": "value"})
    runner.test("Event triggers listener", len(received_data) == 1)
    runner.test("Listener receives correct data", received_data[0] == {"key": "value"})
    
    # Test multiple emissions
    emitter.emit("test.event", "second")
    emitter.emit("test.event", "third")
    runner.test("Multiple emissions work", len(received_data) == 3)
    
    # Test listener count
    runner.test("has_listener returns True", emitter.has_listener("test.event"))
    runner.test("has_listener returns False for unknown", not emitter.has_listener("unknown"))


def run_once_tests(runner: TestRunner):
    """Test once-only listeners."""
    print("\nOnce-Only Listener Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    call_count = [0]
    
    def once_handler(data):
        call_count[0] += 1
    
    emitter.once("once.event", once_handler)
    
    emitter.emit("once.event", "first")
    emitter.emit("once.event", "second")
    emitter.emit("once.event", "third")
    
    runner.test("Once listener called once", call_count[0] == 1)
    runner.test("Listener removed after call", emitter.get_listener_count("once.event") == 0)


def run_wildcard_tests(runner: TestRunner):
    """Test wildcard pattern matching."""
    print("\nWildcard Pattern Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    events = []
    
    def single_wildcard(event, data):
        events.append(("single", event, data))
    
    def double_wildcard(event, data):
        events.append(("double", event, data))
    
    emitter.on("user.*", single_wildcard)
    emitter.on("data.**", double_wildcard)
    
    # Single wildcard - matches one level
    emitter.emit("user.created", {"id": 1})
    emitter.emit("user.deleted", {"id": 2})
    emitter.emit("user.profile.updated", {"id": 3})  # Should NOT match single *
    
    runner.test("Single * matches one level", len([e for e in events if e[0] == "single"]) == 2)
    
    # Double wildcard - matches all levels
    emitter.emit("data.users.sync", {"count": 100})
    emitter.emit("data.config.changed", {"key": "value"})
    
    double_events = [e for e in events if e[0] == "double"]
    runner.test("Double ** matches all levels", len(double_events) == 2)
    
    # Test exact match priority
    exact_called = [False]
    emitter.on("exact.event", lambda d: exact_called.__setitem__(0, True))
    emitter.emit("exact.event")
    runner.test("Exact match works", exact_called[0])


def run_priority_tests(runner: TestRunner):
    """Test listener priority ordering."""
    print("\nPriority Ordering Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    order = []
    
    emitter.on("priority.test", lambda d: order.append(3), priority=0)
    emitter.on("priority.test", lambda d: order.append(1), priority=10)
    emitter.on("priority.test", lambda d: order.append(2), priority=5)
    
    emitter.emit("priority.test")
    
    runner.test("High priority called first", order == [1, 2, 3])


def run_removal_tests(runner: TestRunner):
    """Test listener removal."""
    print("\nListener Removal Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    calls = {"a": 0, "b": 0}
    
    def handler_a(data):
        calls["a"] += 1
    
    def handler_b(data):
        calls["b"] += 1
    
    emitter.on("remove.test", handler_a)
    emitter.on("remove.test", handler_b)
    
    emitter.emit("remove.test")
    runner.test("Both handlers called initially", calls["a"] == 1 and calls["b"] == 1)
    
    # Remove specific handler
    emitter.off("remove.test", handler_a)
    calls["a"] = 0
    calls["b"] = 0
    emitter.emit("remove.test")
    runner.test("Specific handler removed", calls["a"] == 0 and calls["b"] == 1)
    
    # Remove all handlers
    emitter.off("remove.test")
    runner.test("All handlers removed", emitter.get_listener_count("remove.test") == 0)
    
    # Remove all events
    emitter.on("event1", lambda: None)
    emitter.on("event2", lambda: None)
    emitter.off("*")
    runner.test("Wildcard off removes all", emitter.listener_count == 0)


def run_history_tests(runner: TestRunner):
    """Test event history and replay."""
    print("\nEvent History Tests")
    print("-" * 60)
    
    emitter = EventEmitter(max_history=5)
    
    for i in range(10):
        emitter.emit("log.event", f"message_{i}")
    
    history = emitter.get_history("log.event", limit=10)
    runner.test("History limited to max_history", len(history) == 5)
    runner.test("History contains latest events", history[0].data == "message_5")
    
    # Test replay
    replayed = []
    emitter.replay("log.event", lambda d: replayed.append(d))
    runner.test("Replay replays all history", len(replayed) == 5)
    
    # Test clear history
    emitter.clear_history("log.event")
    runner.test("Clear history works", len(emitter.get_history("log.event")) == 0)


def run_pause_resume_tests(runner: TestRunner):
    """Test event pause/resume functionality."""
    print("\nPause/Resume Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    called = [False]
    
    emitter.on("paused.event", lambda d: called.__setitem__(0, True))
    
    emitter.pause("paused.event")
    runner.test("Event is paused", emitter.is_paused("paused.event"))
    
    emitter.emit("paused.event")
    runner.test("Paused event not triggered", not called[0])
    
    emitter.resume("paused.event")
    runner.test("Event is resumed", not emitter.is_paused("paused.event"))
    
    emitter.emit("paused.event")
    runner.test("Resumed event triggered", called[0])


def run_introspection_tests(runner: TestRunner):
    """Test emitter introspection."""
    print("\nIntrospection Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    
    emitter.on("event1", lambda: None, priority=5)
    emitter.on("event1", lambda: None, priority=10)
    emitter.on("event2", lambda: None)
    
    runner.test("get_events returns all events", len(emitter.get_events()) == 2)
    runner.test("get_listener_count correct", emitter.get_listener_count("event1") == 2)
    runner.test("listener_count property correct", emitter.listener_count == 3)
    runner.test("event_count property correct", emitter.event_count == 2)
    
    stats = emitter.get_stats()
    runner.test("Stats has required fields", 
                all(k in stats for k in ["id", "event_count", "total_listeners"]))


def run_channel_tests(runner: TestRunner):
    """Test channel creation."""
    print("\nChannel Tests")
    print("-" * 60)
    
    channel = create_channel("user")
    called = [False]
    
    # Channel should prefix events
    channel.on("created", lambda d: called.__setitem__(0, True))
    channel.emit("created", {"id": 1})
    
    runner.test("Channel prefixes events", called[0])


def run_global_bus_tests(runner: TestRunner):
    """Test global event bus singleton."""
    print("\nGlobal Event Bus Tests")
    print("-" * 60)
    
    # Reset singleton for clean test
    EventBus.reset()
    
    bus1 = EventBus.get_instance()
    bus2 = EventBus.get_instance()
    
    runner.test("Bus is singleton", bus1 is bus2)
    
    called = [False]
    bus1.on("singleton.test", lambda d: called.__setitem__(0, True))
    bus2.emit("singleton.test")
    
    runner.test("Singleton shares state", called[0])
    
    # Cleanup
    EventBus.reset()


def run_decorator_tests(runner: TestRunner):
    """Test decorator utilities."""
    print("\nDecorator Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    called = [False]
    
    @on_event(emitter, "decorated.event")
    def decorated_handler(data):
        called[0] = True
    
    emitter.emit("decorated.event")
    runner.test("on_event decorator registers listener", called[0])


def run_thread_safety_tests(runner: TestRunner):
    """Test thread safety."""
    print("\nThread Safety Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    counter = [0]
    lock = threading.Lock()
    
    def thread_handler(data):
        with lock:
            counter[0] += 1
    
    # Register listener
    emitter.on("thread.test", thread_handler)
    
    # Create multiple threads
    threads = []
    for _ in range(10):
        t = threading.Thread(target=lambda: emitter.emit("thread.test"))
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    runner.test("Thread-safe emission", counter[0] == 10)


def run_async_tests(runner: TestRunner):
    """Test async listener support."""
    print("\nAsync Listener Tests")
    print("-" * 60)
    
    # Skip if asyncio.run not available (Python < 3.7)
    if not hasattr(asyncio, 'run'):
        runner.test("Async support (skipped - Python < 3.7)", True)
        return
    
    emitter = EventEmitter()
    called = [False]
    
    async def async_handler(data):
        called[0] = True
    
    emitter.on("async.test", async_handler)
    
    # Run async event loop
    async def run_test():
        emitter.emit_async("async.test")
        await asyncio.sleep(0.1)  # Give time for async handler
    
    asyncio.run(run_test())
    
    runner.test("Async listener called", called[0])


def run_edge_cases(runner: TestRunner):
    """Test edge cases."""
    print("\nEdge Case Tests")
    print("-" * 60)
    
    emitter = EventEmitter()
    
    # Empty emitter
    runner.test("Empty emitter has 0 listeners", emitter.listener_count == 0)
    runner.test("Empty emitter has 0 events", emitter.event_count == 0)
    
    # Emit with no listeners (should not crash)
    try:
        emitter.emit("no.listeners")
        runner.test("Emit with no listeners doesn't crash", True)
    except Exception as e:
        runner.test("Emit with no listeners doesn't crash", False, str(e))
    
    # None data
    received = []
    emitter.on("none.data", lambda d: received.append(d))
    emitter.emit("none.data", None)
    runner.test("None data handled", received == [None])
    
    # Clear all
    emitter.on("clear1", lambda: None)
    emitter.on("clear2", lambda: None)
    emitter.clear()
    runner.test("Clear removes all", emitter.listener_count == 0)


def run_tests():
    """Run all tests."""
    runner = TestRunner()
    
    run_basic_tests(runner)
    run_once_tests(runner)
    run_wildcard_tests(runner)
    run_priority_tests(runner)
    run_removal_tests(runner)
    run_history_tests(runner)
    run_pause_resume_tests(runner)
    run_introspection_tests(runner)
    run_channel_tests(runner)
    run_global_bus_tests(runner)
    run_decorator_tests(runner)
    run_thread_safety_tests(runner)
    run_async_tests(runner)
    run_edge_cases(runner)
    
    return runner.report()


if __name__ == "__main__":
    print("=" * 60)
    print("Event Emitter Utilities - Test Suite")
    print("=" * 60)
    
    success = run_tests()
    sys.exit(0 if success else 1)

File: JavaScript/event_emitter_utils/event_emitter_utils_test.js
/**
 * EventEmitter Utils Test Suite
 * 
 * Comprehensive tests for the EventEmitter utility
 */

const assert = require('assert');
const { EventEmitter, NamespacedEmitter, createEventEmitter, mixinEventEmitter } = require('./mod.js');

// Test helper
let testCount = 0;
let passCount = 0;
let failCount = 0;

function test(name, fn) {
  testCount++;
  try {
    fn();
    passCount++;
    console.log(`✓ ${name}`);
  } catch (error) {
    failCount++;
    console.log(`✗ ${name}`);
    console.log(`  Error: ${error.message}`);
  }
}

function suite(name, fn) {
  console.log(`\n${name}`);
  console.log('='.repeat(50));
  fn();
}

// ============================================
// Basic EventEmitter Tests
// ============================================

suite('Basic EventEmitter', () => {
  test('should create an EventEmitter instance', () => {
    const emitter = new EventEmitter();
    assert.ok(emitter instanceof EventEmitter);
  });

  test('should subscribe and emit events', () => {
    const emitter = new EventEmitter();
    let called = false;
    
    emitter.on('test', () => {
      called = true;
    });
    
    emitter.emit('test');
    assert.strictEqual(called, true);
  });

  test('should pass arguments to listeners', () => {
    const emitter = new EventEmitter();
    let received = null;
    
    emitter.on('test', (arg) => {
      received = arg;
    });
    
    emitter.emit('test', 'hello');
    assert.strictEqual(received, 'hello');
  });

  test('should pass multiple arguments to listeners', () => {
    const emitter = new EventEmitter();
    let received = null;
    
    emitter.on('test', (a, b, c) => {
      received = [a, b, c];
    });
    
    emitter.emit('test', 1, 2, 3);
    assert.deepStrictEqual(received, [1, 2, 3]);
  });

  test('should return true when event has listeners', () => {
    const emitter = new EventEmitter();
    emitter.on('test', () => {});
    
    const result = emitter.emit('test');
    assert.strictEqual(result, true);
  });

  test('should return false when event has no listeners', () => {
    const emitter = new EventEmitter();
    
    const result = emitter.emit('nonexistent');
    assert.strictEqual(result, false);
  });
});

// ============================================
// once() Tests
// ============================================

suite('once() method', () => {
  test('should fire only once', () => {
    const emitter = new EventEmitter();
    let count = 0;
    
    emitter.once('test', () => {
      count++;
    });
    
    emitter.emit('test');
    emitter.emit('test');
    
    assert.strictEqual(count, 1);
  });

  test('should receive correct arguments', () => {
    const emitter = new EventEmitter();
    let received = null;
    
    emitter.once('test', (arg) => {
      received = arg;
    });
    
    emitter.emit('test', 'data');
    assert.strictEqual(received, 'data');
  });

  test('should remove listener after first emit', () => {
    const emitter = new EventEmitter();
    
    emitter.once('test', () => {});
    
    assert.strictEqual(emitter.listenerCount('test'), 1);
    emitter.emit('test');
    assert.strictEqual(emitter.listenerCount('test'), 0);
  });
});

// ============================================
// off() Tests
// ============================================

suite('off() method', () => {
  test('should remove a specific listener', () => {
    const emitter = new EventEmitter();
    const handler = () => {};
    
    emitter.on('test', handler);
    assert.strictEqual(emitter.listenerCount('test'), 1);
    
    emitter.off('test', handler);
    assert.strictEqual(emitter.listenerCount('test'), 0);
  });

  test('should not remove other listeners', () => {
    const emitter = new EventEmitter();
    const handler1 = () => {};
    const handler2 = () => {};
    
    emitter.on('test', handler1);
    emitter.on('test', handler2);
    
    emitter.off('test', handler1);
    
    assert.strictEqual(emitter.listenerCount('test'), 1);
  });

  test('should do nothing for nonexistent event', () => {
    const emitter = new EventEmitter();
    
    // Should not throw
    emitter.off('nonexistent', () => {});
  });

  test('should return emitter for chaining', () => {
    const emitter = new EventEmitter();
    const result = emitter.off('test', () => {});
    
    assert.strictEqual(result, emitter);
  });
});

// ============================================
// removeAllListeners() Tests
// ============================================

suite('removeAllListeners() method', () => {
  test('should remove all listeners for specific event', () => {
    const emitter = new EventEmitter();
    
    emitter.on('test', () => {});
    emitter.on('test', () => {});
    emitter.on('other', () => {});
    
    emitter.removeAllListeners('test');
    
    assert.strictEqual(emitter.listenerCount('test'), 0);
    assert.strictEqual(emitter.listenerCount('other'), 1);
  });

  test('should remove all listeners when no event specified', () => {
    const emitter = new EventEmitter();
    
    emitter.on('test1', () => {});
    emitter.on('test2', () => {});
    emitter.on('test3', () => {});
    
    emitter.removeAllListeners();
    
    assert.deepStrictEqual(emitter.eventNames(), []);
  });
});

// ============================================
// Priority Tests
// ============================================

suite('Listener priority', () => {
  test('should call listeners in priority order', () => {
    const emitter = new EventEmitter();
    const order = [];
    
    emitter.on('test', () => order.push('low'), { priority: 1 });
    emitter.on('test', () => order.push('high'), { priority: 10 });
    emitter.on('test', () => order.push('medium'), { priority: 5 });
    
    emitter.emit('test');
    
    assert.deepStrictEqual(order, ['high', 'medium', 'low']);
  });

  test('should handle equal priorities', () => {
    const emitter = new EventEmitter();
    const order = [];
    
    emitter.on('test', () => order.push(1), { priority: 5 });
    emitter.on('test', () => order.push(2), { priority: 5 });
    emitter.on('test', () => order.push(3), { priority: 5 });
    
    emitter.emit('test');
    
    assert.strictEqual(order.length, 3);
  });
});

// ============================================
// Wildcard Tests
// ============================================

suite('Wildcard matching', () => {
  test('should match single segment wildcards', () => {
    const emitter = new EventEmitter();
    let matched = false;
    
    emitter.on('user.*', () => {
      matched = true;
    });
    
    emitter.emit('user.created');
    assert.strictEqual(matched, true);
  });

  test('should match multiple segment wildcards', () => {
    const emitter = new EventEmitter();
    const matched = [];
    
    emitter.on('user.**', () => {
      matched.push(true);
    });
    
    emitter.emit('user.created');
    emitter.emit('user.profile.updated');
    emitter.emit('user.settings.theme.changed');
    
    assert.strictEqual(matched.length, 3);
  });

  test('should not match incorrect patterns', () => {
    const emitter = new EventEmitter();
    let matched = false;
    
    emitter.on('user.created', () => {
      matched = true;
    });
    
    emitter.emit('user.updated');
    assert.strictEqual(matched, false);
  });
});

// ============================================
// Async Tests
// ============================================

suite('Async emit', () => {
  test('should handle async emit', async () => {
    const emitter = new EventEmitter();
    let called = false;
    
    emitter.on('test', async () => {
      await new Promise(resolve => setTimeout(resolve, 10));
      called = true;
    });
    
    await emitter.emitAsync('test');
    assert.strictEqual(called, true);
  });

  test('should handle sync listeners in async emit', async () => {
    const emitter = new EventEmitter();
    let called = false;
    
    emitter.on('test', () => {
      called = true;
    });
    
    await emitter.emitAsync('test');
    assert.strictEqual(called, true);
  });

  test('should handle mixed sync and async listeners', async () => {
    const emitter = new EventEmitter();
    const order = [];
    
    emitter.on('test', () => order.push(1));
    emitter.on('test', async () => {
      await new Promise(resolve => setTimeout(resolve, 5));
      order.push(2);
    });
    emitter.on('test', () => order.push(3));
    
    await emitter.emitAsync('test');
    
    assert.deepStrictEqual(order, [1, 2, 3]);
  });
});

// ============================================
// History Tests
// ============================================

suite('Event history', () => {
  test('should capture history when enabled', () => {
    const emitter = new EventEmitter({ captureHistory: true });
    
    emitter.emit('test', 'data1');
    emitter.emit('test', 'data2');
    emitter.emit('other', 'data3');
    
    const history = emitter.getHistory();
    
    assert.strictEqual(history.length, 3);
    assert.strictEqual(history[0].event, 'test');
    assert.deepStrictEqual(history[0].args, ['data1']);
  });

  test('should not capture history when disabled', () => {
    const emitter = new EventEmitter({ captureHistory: false });
    
    emitter.emit('test', 'data');
    
    const history = emitter.getHistory();
    assert.strictEqual(history.length, 0);
  });

  test('should filter history by event name', () => {
    const emitter = new EventEmitter({ captureHistory: true });
    
    emitter.emit('test', 'data1');
    emitter.emit('other', 'data2');
    emitter.emit('test', 'data3');
    
    const history = emitter.getHistory('test');
    assert.strictEqual(history.length, 2);
  });

  test('should limit history size', () => {
    const emitter = new EventEmitter({ captureHistory: true, historySize: 5 });
    
    for (let i = 0; i < 10; i++) {
      emitter.emit('test', i);
    }
    
    const history = emitter.getHistory();
    assert.strictEqual(history.length, 5);
    assert.strictEqual(history[0].args[0], 5); // First kept event
    assert.strictEqual(history[4].args[0], 9); // Last event
  });

  test('should replay historical event', () => {
    const emitter = new EventEmitter({ captureHistory: true });
    const results = [];
    
    emitter.on('test', (data) => results.push(data));
    
    emitter.emit('test', 'original');
    
    // Remove the listener and add a new one
    emitter.removeAllListeners('test');
    emitter.on('test', (data) => results.push(`replayed: ${data}`));
    
    emitter.replay(0);
    
    assert.deepStrictEqual(results, ['original', 'replayed: original']);
  });

  test('should clear history', () => {
    const emitter = new EventEmitter({ captureHistory: true });
    
    emitter.emit('test', 'data');
    emitter.clearHistory();
    
    const history = emitter.getHistory();
    assert.strictEqual(history.length, 0);
  });
});

// ============================================
// Namespace Tests
// ============================================

suite('Namespaced emitter', () => {
  test('should prefix event names with namespace', () => {
    const emitter = new EventEmitter();
    const ns = emitter.namespace('user');
    let received = null;
    
    ns.on('created', (data) => {
      received = data;
    });
    
    ns.emit('created', 'test data');
    
    assert.strictEqual(received, 'test data');
    assert.strictEqual(emitter.listenerCount('user.created'), 1);
  });

  test('should work with once', () => {
    const emitter = new EventEmitter();
    const ns = emitter.namespace('app');
    let count = 0;
    
    ns.once('init', () => count++);
    
    ns.emit('init');
    ns.emit('init');
    
    assert.strictEqual(count, 1);
  });

  test('should work with off', () => {
    const emitter = new EventEmitter();
    const ns = emitter.namespace('app');
    const handler = () => {};
    
    ns.on('test', handler);
    assert.strictEqual(ns.listenerCount('test'), 1);
    
    ns.off('test', handler);
    assert.strictEqual(ns.listenerCount('test'), 0);
  });
});

// ============================================
// Mixin Tests
// ============================================

suite('mixinEventEmitter', () => {
  test('should add EventEmitter methods to object', () => {
    const obj = {};
    mixinEventEmitter(obj);
    
    assert.ok(typeof obj.on === 'function');
    assert.ok(typeof obj.off === 'function');
    assert.ok(typeof obj.emit === 'function');
    assert.ok(typeof obj.once === 'function');
  });

  test('should allow object to emit events', () => {
    const obj = { name: 'test' };
    mixinEventEmitter(obj);
    
    let received = null;
    obj.on('change', (data) => {
      received = data;
    });
    
    obj.emit('change', 'updated');
    
    assert.strictEqual(received, 'updated');
  });
});

// ============================================
// createEventEmitter Tests
// ============================================

suite('createEventEmitter', () => {
  test('should create EventEmitter instance', () => {
    const emitter = createEventEmitter();
    assert.ok(emitter instanceof EventEmitter);
  });

  test('should accept options', () => {
    const emitter = createEventEmitter({ maxListeners: 50 });
    assert.strictEqual(emitter.getMaxListeners(), 50);
  });
});

// ============================================
// Error Handling Tests
// ============================================

suite('Error handling', () => {
  test('should handle errors in listeners gracefully', () => {
    const emitter = new EventEmitter();
    let secondCalled = false;
    
    emitter.on('test', () => {
      throw new Error('Test error');
    });
    
    emitter.on('test', () => {
      secondCalled = true;
    });
    
    // Should not throw
    emitter.emit('test');
    
    // Second listener should still be called
    assert.strictEqual(secondCalled, true);
  });

  test('should throw for non-function listener', () => {
    const emitter = new EventEmitter();
    
    assert.throws(() => {
      emitter.on('test', 'not a function');
    }, TypeError);
  });

  test('should throw for non-function once listener', () => {
    const emitter = new EventEmitter();
    
    assert.throws(() => {
      emitter.once('test', {});
    }, TypeError);
  });
});

// ============================================
// Max Listeners Tests
// ============================================

suite('Max listeners', () => {
  test('should warn when exceeding max listeners', () => {
    const warnings = [];
    const originalWarn = console.warn;
    console.warn = (msg) => warnings.push(msg);
    
    const emitter = new EventEmitter({ maxListeners: 2 });
    
    emitter.on('test', () => {});
    emitter.on('test', () => {});
    emitter.on('test', () => {}); // Should trigger warning
    
    console.warn = originalWarn;
    
    assert.strictEqual(warnings.length, 1);
    assert.ok(warnings[0].includes('memory leak'));
  });

  test('should allow setting max listeners', () => {
    const emitter = new EventEmitter();
    
    emitter.setMaxListeners(50);
    
    assert.strictEqual(emitter.getMaxListeners(), 50);
  });
});

// ============================================
// Edge Cases
// ============================================

suite('Edge cases', () => {
  test('should handle removing listener during iteration', () => {
    const emitter = new EventEmitter();
    const order = [];
    
    const handler1 = () => {
      order.push(1);
      emitter.off('test', handler2);
    };
    const handler2 = () => order.push(2);
    
    emitter.on('test', handler1);
    emitter.on('test', handler2);
    
    emitter.emit('test');
    
    assert.deepStrictEqual(order, [1, 2]);
  });

  test('should handle self-removing listener', () => {
    const emitter = new EventEmitter();
    const handler = () => {
      emitter.off('test', handler);
    };
    
    emitter.on('test', handler);
    
    emitter.emit('test');
    emitter.emit('test');
    
    assert.strictEqual(emitter.listenerCount('test'), 0);
  });

  test('should return unsubscribe function from on()', () => {
    const emitter = new EventEmitter();
    let count = 0;
    
    const unsubscribe = emitter.on('test', () => count++);
    
    emitter.emit('test');
    assert.strictEqual(count, 1);
    
    unsubscribe();
    emitter.emit('test');
    assert.strictEqual(count, 1); // Still 1, listener removed
  });

  test('should return unsubscribe function from once()', () => {
    const emitter = new EventEmitter();
    let count = 0;
    
    const unsubscribe = emitter.once('test', () => count++);
    
    unsubscribe();
    emitter.emit('test');
    
    assert.strictEqual(count, 0);
  });

  test('should handle empty event names', () => {
    const emitter = new EventEmitter();
    let called = false;
    
    emitter.on('', () => called = true);
    emitter.emit('');
    
    assert.strictEqual(called, true);
  });

  test('should handle special characters in event names', () => {
    const emitter = new EventEmitter();
    let called = false;
    
    emitter.on('user:action/created', () => called = true);
    emitter.emit('user:action/created');
    
    assert.strictEqual(called, true);
  });
});

// ============================================
// Performance Tests
// ============================================

suite('Performance', () => {
  test('should handle many listeners efficiently', () => {
    const emitter = new EventEmitter({ maxListeners: 10000 });
    const start = Date.now();
    
    for (let i = 0; i < 1000; i++) {
      emitter.on('test', () => {});
    }
    
    const elapsed = Date.now() - start;
    assert.ok(elapsed < 100, `Adding 1000 listeners took ${elapsed}ms`);
    
    emitter.removeAllListeners('test');
    assert.strictEqual(emitter.listenerCount('test'), 0);
  });

  test('should emit to many listeners efficiently', () => {
    const emitter = new EventEmitter({ maxListeners: 10000 });
    let count = 0;
    
    for (let i = 0; i < 100; i++) {
      emitter.on('test', () => count++);
    }
    
    const start = Date.now();
    emitter.emit('test');
    const elapsed = Date.now() - start;
    
    assert.strictEqual(count, 100);
    assert.ok(elapsed < 10, `Emitting to 100 listeners took ${elapsed}ms`);
  });
});

// ============================================
// Print Summary
// ============================================

console.log('\n' + '='.repeat(50));
console.log(`Tests: ${testCount}`);
console.log(`Passed: ${passCount}`);
console.log(`Failed: ${failCount}`);
console.log('='.repeat(50));

if (failCount === 0) {
  console.log('\n✓ All tests passed!\n');
  process.exit(0);
} else {
  console.log('\n✗ Some tests failed.\n');
  process.exit(1);
}